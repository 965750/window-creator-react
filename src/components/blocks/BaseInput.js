import React, { Component } from 'react'
import styles from './BaseInput.module.scss'
import { injectIntl, FormattedMessage } from 'react-intl'
import PropTypes from 'prop-types'

class BaseInput extends Component {
    state = {
        error: null,
    }

    validate = () => {
        this.setState({
            error: null,
        })

        if (this.props.toValide.indexOf('email') > -1) {
            const emailCheck = /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/

            if (!emailCheck.test(this.props.value)) {
                this.setState({
                    error: 'Invalid email address',
                })
            }
        }

        if (this.props.toValide.indexOf('isRequired') > -1) {
            if (this.props.value.length <= 0) {
                this.setState({
                    error: 'Field is required',
                })
            }
        }

        if (this.props.toValide.indexOf('match') > -1) {
            if (this.props.value !== this.props.valueToMatch) {
                this.setState({
                    error: 'Passwords are not equal',
                })
            }
        }

        if (this.props.toValide.indexOf('passwordLength') > -1) {
            if (this.props.value.length <= 7) {
                this.setState({
                    error: 'Password is too short',
                })
            }
        }
    }
    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.value !== this.props.value) {
            this.validate()
        }

        if (this.props.valueToMatch && this.props.value) {
            if (
                prevProps.value !== this.props.value ||
                prevProps.valueToMatch !== this.props.valueToMatch
            ) {
                this.validate()
            }
        }

        if (prevState.error !== this.state.error) {
            this.props.validationError(this.state.error)
        }
    }

    render() {
        let errorMsg

        if (this.state.error)
            errorMsg = (
                <p className={styles.field__errorMsg}>
                    <FormattedMessage
                        id={this.state.error}
                        defaultMessage={this.state.error}
                    />
                </p>
            )

        return (
            <div
                className={`${styles.field__wrapper} ${
                    this.props.classes ? this.props.classes : ''
                }`}
            >
                <input
                    id={this.props.id}
                    onChange={this.props.handleChange}
                    value={this.props.value}
                    type={this.props.type}
                    placeholder={this.props.intl.formatMessage({
                        id: this.props.placeholder,
                    })}
                    className={`${styles.field} ${
                        this.state.error ? styles['field--error'] : ''
                    } border border-light rounded p-3 w-full`}
                />
                {errorMsg}
            </div>
        )
    }
}

BaseInput.propTypes = {
    handleChange: PropTypes.func,
    validationError: PropTypes.func,
    value: PropTypes.string,
    placeholder: PropTypes.string,
    toValide: PropTypes.array,
    classes: PropTypes.string,
    type: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
}

export default injectIntl(BaseInput)
