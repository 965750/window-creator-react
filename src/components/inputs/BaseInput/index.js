import { FormattedMessage, injectIntl } from 'react-intl'

import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './BaseInput.module.scss'

class BaseInput extends Component {
  state = {
    error: null,
  }

  componentDidMount() {
    if (this.props.value) {
      this.validate()
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.value !== this.props.value) {
      this.validate()
    }

    if (this.props.valueToMatch && this.props.value) {
      if (
        prevProps.value !== this.props.value
        || prevProps.valueToMatch !== this.props.valueToMatch
      ) {
        this.validate()
      }
    }

    if (prevState.error !== this.state.error) {
      this.props.validationError(this.state.error)
    }
  }

  validate = () => {
    this.setState({
      error: null,
    })

    if (this.props.toValidate.indexOf('email') > -1) {
      const emailCheck = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

      if (!emailCheck.test(this.props.value)) {
        this.setState({
          error: 'Invalid email address',
        })
      }
    }

    if (this.props.toValidate.indexOf('isRequired') > -1) {
      if (this.props.value.length <= 0) {
        this.setState({
          error: 'Field is required',
        })
      }
    }

    if (this.props.toValidate.indexOf('match') > -1) {
      if (this.props.value !== this.props.valueToMatch) {
        this.setState({
          error: 'Passwords are not equal',
        })
      }
    }

    if (this.props.toValidate.indexOf('passwordLength') > -1) {
      if (this.props.value.length <= 7) {
        this.setState({
          error: 'Password is too short',
        })
      }
    }

    if (this.props.toValidate.indexOf('tooLong') > -1) {
      if (this.props.value.length >= 12) {
        this.setState({
          error: 'Name must be shorter',
        })
      }
    }
  }

  render() {
    let errorMsg

    if (this.state.error) {
      errorMsg = (
        <p className={styles.field__errorMsg}>
          <FormattedMessage
            id={this.state.error}
            defaultMessage={this.state.error}
          />
        </p>
      )
    }

    return (
      <div
        data-testid="inputWrapper"
        className={`${styles.field__wrapper} ${
          this.props.classes ? this.props.classes : ''
        }`}
      >
        <input
          id={this.props.id}
          onChange={this.props.handleChange}
          value={this.props.value}
          type={this.props.type}
          placeholder={this.props.intl.formatMessage({
            id: this.props.placeholder,
          })}
          className={`${styles.field} ${
            this.state.error ? styles['field--error'] : ''
          } border border-silver rounded p-3 w-full`}
        />
        {errorMsg}
      </div>
    )
  }
}

BaseInput.propTypes = {
  handleChange: PropTypes.func,
  validationError: PropTypes.func,
  value: PropTypes.string,
  placeholder: PropTypes.string,
  toValidate: PropTypes.arrayOf(PropTypes.string),
  intl: PropTypes.shape({
    formatMessage: PropTypes.func,
  }),
  valueToMatch: PropTypes.string,
  classes: PropTypes.string,
  type: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
}

export default injectIntl(BaseInput)
