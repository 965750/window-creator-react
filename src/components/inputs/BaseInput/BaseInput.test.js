import '@testing-library/jest-dom/extend-expect'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup } from '@testing-library/react'

import BaseInput from './index'
import messages from '../../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
  },
}

const reducer = (state = initialStore, action) => {
  switch (action.type) {
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('BaseInput', () => {
  let wrapper
  const initialProps = {
    handleChange: jest.fn(),
    validationError: jest.fn(),
    value: '',
    placeholder: 'First name',
    id: 'mockId',
    type: 'text',
    classes: '',
    valueToMatch: '',
    toValidate: ['']
  }

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale="en"
        messages={messages[store.getState().theme.lang]}
      >
        <BaseInput {...props} />
      </IntlProvider>,
    )
  }

  afterEach(cleanup)

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should check validation for too long value', () => {
    setUp({
      ...initialProps,
      toValidate: ['tooLong'],
      value: '123456789',
    })

    expect(wrapper.queryByText(/Nazwa musi być krótsza/)).toBeNull()

    cleanup()

    setUp({
      ...initialProps,
      toValidate: ['tooLong'],
      value: '123456789asdf',
    })

    expect(wrapper.queryByText(/Nazwa musi być krótsza/)).not.toBeNull()
  })

  it('should check validation for too short password', () => {
    setUp({
      ...initialProps,
      toValidate: ['passwordLength'],
      value: '123456789',
    })

    expect(wrapper.queryByText(/Hasło jest za krótkie/)).toBeNull()

    cleanup()

    setUp({
      ...initialProps,
      toValidate: ['passwordLength'],
      type: 'password',
      value: '123456',
    })

    expect(wrapper.queryByText(/Hasło jest za krótkie/)).not.toBeNull()
  })

  it('should check validation for matching password', () => {
    setUp({
      ...initialProps,
      toValidate: ['match'],
      value: '12345678',
      valueToMatch: '12345678',
    })

    expect(wrapper.queryByText(/Hasła nie są identyczne/)).toBeNull()

    cleanup()

    setUp({
      ...initialProps,
      toValidate: ['match'],
      type: 'password',
      value: '12345678',
      valueToMatch: '1234567'
    })

    expect(wrapper.queryByText(/Hasła nie są identyczne/)).not.toBeNull()
  })

  it('should check validation for valid email address', () => {
    setUp({
      ...initialProps,
      toValidate: ['email'],
      value: 'email@gmail.com'
    })

    expect(wrapper.queryByText(/Niewłaściwy adres email/)).toBeNull()

    cleanup()

    setUp({
      ...initialProps,
      toValidate: ['email'],
      type: 'text',
      value: 'email@emial',
    })

    expect(wrapper.queryByText(/Niewłaściwy adres email/)).not.toBeNull()
  })

  it('should pass classes from props to wrapper', () => {
    const mockClasses = 'mockClasses'

    setUp({
      ...initialProps,
      type: 'text',
      classes: mockClasses
    })

    expect(wrapper.queryByTestId('inputWrapper')).toHaveClass(mockClasses)
  })
})
