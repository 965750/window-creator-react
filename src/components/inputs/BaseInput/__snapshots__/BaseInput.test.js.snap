// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="field__wrapper "
    data-testid="inputWrapper"
  >
    <input
      class="field  border border-silver rounded p-3 w-full"
      id="mockId"
      placeholder="Imie"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
